### ISSUE TITLE
Pending 'missing' case issue re btrfs fi show

### ISSUE THREAD

**Issue by phillxnet** [rockstor/rockstor-core#2311]
There is a pending proposal to enhance the output of 'btrfs fi show' to include the devid and an advisory, or missing, device path. I.e. currently there is the following expected output of btrfs fi show:

```
...
        devid    1 size 5.00GiB used 1.26GiB path /dev/loop0
        *** Some devices missing
```
proposed to be replaced with:
```
...
        devid    1 size 5.00GiB used 1.26GiB path /dev/loop0
        devid    2 size 0 used 0 path /dev/loop1 MISSING
```

Where the device path "/dev/loop1" in this case, may be omitted if unavailable.

This additional missing dev info could be of great use to our 'missing' logic which is currently a little convoluted (for another issue) but more pressing is the capitalisation of the "missing" to "MISSING" as we currently have  an overarching flaging mechanism that simply searches for any line ending with "missing" in 'btrfs fi show':

See in fs/btrfs:
- is_pool_missing_dev(label)
- degraded_pools_found()

See also our test data for examples of captured 'btrfs fi show' output with pools where a device has one or more missing members:
- fs/tests/test_btrfs.py test_degraded_pools_found(self)

btrfs mailing list references:
https://lore.kernel.org/linux-btrfs/f31fb924-847a-8b54-3da6-707914135f05@suse.com/T/#md987208db4b27afe89e903d6873e2c5c36f3c167
https://lore.kernel.org/linux-btrfs/b983789e-3671-a5f3-1803-37b8992261b4@suse.com/T/#t

**Comment by phillxnet** [comment]
This case issue in upstream is as yet not seen in the current as of writing 5.15.5 Kernel_stable_Backport repo provided kernel:
See: https://github.com/rockstor/rockstor-installer/pull/87 testing comments.

**Comment by phillxnet** [comment]
We are also still OK on this proposed case change from "missing" to "MISSING" in 5.16.0 rc5 via Kernel_HEAD_Backport repo kernel:
See related kernel repo options pr in our rockstor-installer repo:
https://github.com/rockstor/rockstor-installer/pull/87#issuecomment-992654204

![5 16 0-rc5-pl153 2-missing-disk](https://user-images.githubusercontent.com/2521585/145856433-47a29ac3-dcb7-4c10-818a-addb8b9a7251.png)

```
installer-uefi:~ # btrfs fi show /mnt2/rock-pool
Label: 'rock-pool'  uuid: 765b919f-69de-4cc1-842c-20bba390d0d0
        Total devices 2 FS bytes used 1.85GiB
        devid    2 size 8.00GiB used 2.54GiB path /dev/sdc
        *** Some devices missing
```

**Comment by phillxnet** [comment]
As we don't yet see this change in Kernel_HEAD_Backport kernel version and we are far from this currently with our default, or even Kernel_stable_Backport I think it's best not to second guess the final format and go with what we have currently for our "First 4 Stable (ISO).

As such I don't think we need have this issue in our pending First v4 Stable milestone as I had considered earlier. If a similar change ends up being added then we can release our adaptation as and when to the stable update channel.

Leaving this issue for the time being on the ground of "... if it's not broken then don't fix it.". The indicated change is not seen in the tested 5.16 rc5 Kernel_HEAD_Backport so I think we are OK for a bit. Plus as we have no concrete example kernel output we have no solid test data to prove our 'fix' and guard against regression on this rather important feature.

**Comment by phillxnet** [comment]
As of this comment date, our Leap 15.4 base, with Stable Kernel backport as per linked doc pull request, results in the following kernel version:

```
buildvm:~ # uname -a
Linux buildvm 6.1.8-lp153.4.gbddbb73-default #1 SMP PREEMPT_DYNAMIC Mon Jan 30 06:53:24 UTC 2023 (bddbb73) x86_64 x86_64 x86_64 GNU/Linux
```

```
buildvm:~ # zypper lr -U
Repository priorities in effect:                                                                                                                                                         (See 'zypper lr -P' for details)
      97 (raised priority)  :  1 repository
      99 (default priority) :  7 repositories
     105 (lowered priority) :  1 repository

#  | Alias                              | Name                                                  | Enabled | GPG Check | Refresh | URI
---+------------------------------------+-------------------------------------------------------+---------+-----------+---------+----------------------------------------------------------------------------------------
10 | repo-sle-debug-update              | Update repository with debuginfo for updates from S-> | No      | ----      | ----    | http://download.opensuse.org/debug/update/leap/15.4/sle/
 2 | Leap_15_4                          | Leap_15_4                                             | Yes     | (r ) Yes  | Yes     | http://download.opensuse.org/distribution/leap/15.4/repo/oss/
 9 | repo-backports-update              | Update repository of openSUSE Backports               | Yes     | (r ) Yes  | Yes     | http://download.opensuse.org/update/leap/15.4/backports/
 8 | repo-backports-debug-update        | Update repository with updates for openSUSE Leap de-> | No      | ----      | ----    | http://download.opensuse.org/update/leap/15.4/backports_debug/
11 | repo-sle-update                    | Update repository with updates from SUSE Linux Ente-> | Yes     | (r ) Yes  | Yes     | http://download.opensuse.org/update/leap/15.4/sle/
 4 | Rockstor-Testing                   | Rockstor-Testing                                      | Yes     | (r ) Yes  | Yes     | http://updates.rockstor.com:8999/rockstor-testing/leap/15.4
 1 | Kernel_stable_Backport             | Kernel_stable_Backport                                | Yes     | (r ) Yes  | Yes     | https://download.opensuse.org/repositories/Kernel:/stable:/Backport/standard/
 5 | filesystems                        | filesystems                                           | Yes     | (r ) Yes  | Yes     | https://download.opensuse.org/repositories/filesystems/15.4/
 6 | home_rockstor                      | home_rockstor                                         | Yes     | (r ) Yes  | Yes     | https://download.opensuse.org/repositories/home:/rockstor/15.4/
 7 | home_rockstor_branches_Base_System | home_rockstor_branches_Base_System                    | Yes     | (r ) Yes  | Yes     | https://download.opensuse.org/repositories/home:/rockstor:/branches:/Base:/System/15.4/
 3 | Leap_15_4_Updates                  | Leap_15_4_Updates                                     | Yes     | (r ) Yes  | Yes     | https://download.opensuse.org/update/leap/15.4/oss/

```

**Comment by phillxnet** [comment]
As of this comment date we now have:
```
uname -a
Linux buildvm 6.1.9-lp153.2.g79d6a70-default #1 SMP PREEMPT_DYNAMIC Wed Feb  1 07:45:32 UTC 2023 (79d6a70) x86_64 x86_64 x86_64 GNU/Linux
```

And this now shows the expected behaviour referened in the original text of this issue:
```
btrfs fi show
Label: 'ROOT'  uuid: 9ccfb511-b222-4528-944c-4837b9eb089a
        Total devices 1 FS bytes used 3.05GiB
        devid    1 size 17.93GiB used 3.55GiB path /dev/sdb4

Label: 'test-pool-new-kernel'  uuid: 21345a94-f2bf-48d7-a2be-37734ffd2a48
        Total devices 4 FS bytes used 4.20GiB
        devid    1 size 0 used 0 path  MISSING
        devid    2 size 5.00GiB used 5.00GiB path /dev/sdc
        devid    3 size 5.00GiB used 5.00GiB path /dev/sdd
        devid    4 size 5.00GiB used 20.00MiB path /dev/sde
```
Where we, as expected, fail to surface within the Web-UI this critical condition: currently suspected as bing down to the case of "missing" vs "MISSING".

**Comment by phillxnet** [comment]
After modifying our existing unit test: test_degraded_pools_found() to now include the above reproducer of the
above (prior comment) newer format of output we have the expected failure:

For speed we run only our btrfs lower-level tests:
```
cd /opt/rockstor/src/rockstor/fs/
export DJANGO_SETTINGS_MODULE=settings
poetry run django-admin test -v 2
```

```
test_degraded_pools_found (rockstor.fs.tests.test_btrfs.BTRFSTests) ... FAIL
======================================================================
FAIL: test_degraded_pools_found (rockstor.fs.tests.test_btrfs.BTRFSTests)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/opt/rockstor/src/rockstor/fs/tests/test_btrfs.py", line 1645, in test_degraded_pools_found
    "count expected ({})".format(out, count),
AssertionError: Un-expected degraded pool count. Mock (["Label: 'ROOT'  uuid: 9ccfb511-b222-4528-944c-4837b9eb089a", '\tTotal devices 1 FS bytes used 3275034624', '\tdevid    1 size 19257077760 used 3808428032 path /dev/sdb4', '', "Label: 'test-pool-new-kernel'  uuid: 21345a94-f2bf-48d7-a2be-37734ffd2a48", '\tTotal devices 4 FS bytes used 4508352512', '\tdevid    1 size 0 used 0 path  MISSING', '\tdevid    2 size 5368709120 used 5346689024 path /dev/sdc', '\tdevid    3 size 5368709120 used 5346689024 path /dev/sdd', '\tdevid    4 size 5368709120 used 0 path /dev/sde', '', '']) count expected (1)

----------------------------------------------------------------------
Ran 47 tests in 0.071s

FAILED (failures=1)
```
Our test data has the correct output as 1 failed device (count expected (1)). But due to this case issue no "missing" were found!

**Comment by phillxnet** [comment]
Adapting our existing degraded_pools_found() (the tested code in the last comment) results in all test data now passing:

>test_degraded_pools_found (rockstor.fs.tests.test_btrfs.BTRFSTests) ... ok

And our "Pools Degraded Alert" red flashing Web-UI header warning now appears as expected. But the mouse over of this alert states:"Pools found degraded: (1) unimported" as we have another pool specific 'tell' of degraded status:

is_pool_missing_dev(label):

is also in need of some modification given a minor change in kernel output. A pull request will follow with further tests covering the changes requried to is_pool_missing(), as well as the expanded test data for degraded_pools_found().

**Comment by phillxnet** [comment]
Closing as:
Fixed by #2492
